data/kata-collect-data.sh.in

```diff
@@ -0,0 +1,570 @@
+	for project in @PROJECT_TYPE@
+	do
+		pattern+="|${project}-proxy"
+		pattern+="|${project}-runtime"
+		pattern+="|${project}-shim"
+	done
+
+	# assets
+	pattern+="|clear-containers-image"
+	pattern+="|linux-container"
+
+	# optimised hypervisor
+	pattern+="|qemu-lite"
+
+	# default distro hypervisor
+	pattern+="|qemu-system-x86"
+
+	pattern+=")"
+
+	if have_cmd "dpkg"; then
+		run_cmd_and_show_quoted_output "dpkg -l|egrep \"$pattern\""
+	fi
+
+	if have_cmd "rpm"; then
+		run_cmd_and_show_quoted_output "rpm -qa|egrep \"$pattern\""
+	fi
+
+	separator
+}
+
+show_container_mgr_details()
+{
+	heading "Container manager details"
+
+	if have_cmd "docker"; then
+		subheading "Docker"
+		run_cmd_and_show_quoted_output "docker version"
+		run_cmd_and_show_quoted_output "docker info"
+		run_cmd_and_show_quoted_output "systemctl show docker"
+	fi
+
+	if have_cmd "kubectl"; then
+		subheading "Kubernetes"
+		run_cmd_and_show_quoted_output "kubectl version"
+		run_cmd_and_show_quoted_output "kubectl config view"
+		run_cmd_and_show_quoted_output "systemctl show kubelet"
+
+		if have_cmd "crio"; then
+			run_cmd_and_show_quoted_output "crio --version"
+			run_cmd_and_show_quoted_output "systemctl show crio"
+		fi
+	fi
+
+	separator
+}
+
+show_meta()
+{
+	local date
+
+	heading "Meta details"
+
+	date=$(date '+%Y-%m-%d.%H:%M:%S.%N%z')
+	msg "Running \`$script_name\` version \`$script_version\` at \`$date\`."
+
+	separator
+}
+
+show_runtime()
+{
+	local cmd
+
+	msg "Runtime is \`$runtime\`."
+
+	cmd="@PROJECT_TYPE@-env"
+	heading "\`$cmd\`"
+	run_cmd_and_show_quoted_output "$runtime $cmd"
+
+	separator
+}
+
+# Parameter 1: Path to disk image file.
+# Returns: Agent version string, or "$unknown" on error.
+get_agent_version()
+{
+	local img="$1"
+
+	[ -z "$img" ] && { echo "$unknown"; return;}
+	[ -e "$img" ] || { echo "$unknown"; return;}
+
+	local loop_device
+	local partition_path
+	local partitions
+	local partition
+	local count
+	local mountpoint
+	local version
+	local expected
+
+	loop_device=$(loopmount_image "$img")
+	if [ -z "$loop_device" ]; then
+		echo "$unknown"
+		return
+	fi
+
+	partitions=$(get_partitions "$loop_device")
+	count=$(echo "$partitions"|wc -l)
+
+	expected=1
+
+	if [ "$count" -ne "$expected" ]; then
+		release_device "$loop_device"
+		echo "$unknown"
+		return
+	fi
+
+	partition="$partitions"
+
+	partition_path="/dev/${partition}"
+	if [ ! -e "$partition_path" ]; then
+		release_device "$loop_device"
+		echo "$unknown"
+		return
+	fi
+
+	mountpoint=$(mount_partition "$partition_path")
+
+	agent="/bin/@PROJECT_TYPE@-agent"
+
+	version=$(chroot "$mountpoint" "$agent" --version 2>/dev/null)
+
+	unmount_partition "$mountpoint"
+	release_device "$loop_device"
+
+	[ -z "$version" ] && version="$unknown"
+
+	echo "$version"
+}
+
+# Returns: Full path to the image file.
+get_image_file()
+{
+	local cmd="@PROJECT_TYPE@-env"
+	local cmdline="$runtime $cmd"
+
+	image=$(eval "$cmdline" 2>/dev/null |\
+		grep -A 1 '\[Image\]' |\
```

Comment:
Maybe `'^\[Image\]'` to be sure we pick up that top level section?
The `grep -A 1` is obviously a bit tied to the exact output of the `xxx-env` command - probably not an issue, but could fall over one day.

I did something v.similar a couple of days ago with some awk like:
```
	RUNTIME_PATH=$(awk '/^\[Runtime\]$/ {foundit=1} /^    Path =/ { if (foundit==1) {print $3; foundit=0} } ' <<< "$toml" | sed 's/"//g')
```

just for reference. Also not perfect I think.

Reply:
I've added the anchor and I agree about the fagility issue, but until we start producing output in a superior machine readable format (JSON, but ideally YAML), I think we're going to continue to have this risk.

Did you mention you were going to add JSON support to `kata-env`... ? :smile: